Agents/HistoricalPumpDataInstantiation/HistoricalPumpDataInstantiationAgent/src/main/java/uk/ac/cam/cares/jps/agent/TimeSeriesClientDecorator.java

```diff
@@ -0,0 +1,173 @@
+        if (timeSeriesDoesNotExist(iris)) {
+            LOGGER.debug("Time series does not exist on RDB. Generating time series templates...");
+            // Get the classes (datatype) corresponding to each measure's name needed for initialization
+            // excelReadings.get(header) returns array list, so add get method to get the value
+            List<Class<?>> classes = excelHeaders.stream()
+                    .map(header -> excelReadings.get(header).get(0).getClass())
+                    .collect(Collectors.toList());
+            try (Connection conn = rdbClient.getConnection()) {
+                tsClient.initTimeSeries(iris, classes, this.timeUnit, conn, TimeSeriesClient.Type.STEPWISECUMULATIVE, null, null);
+                LOGGER.debug(String.format("Initialized time series with the following IRIs: %s", String.join(", ", iris)));
+            } finally {
+                // Always close connection
+                rdbClient.getConnection().close();
+            }
+        }
+    }
+
+    /**
+     * Checks if a time series does not exist in the central RDB lookup table using the time series client.
+     * This is achieved through ensuring that all the IRIs are linked and initialised to a time series.
+     *
+     * @param iris A list of string containing IRIs that should be attached to the same time series.
+     * @return False if any IRI have a time series attached, true otherwise.
+     */
+    private boolean timeSeriesDoesNotExist(List<String> iris) throws SQLException {
+        // If any of the IRIs does not have a time series the time series does not exist
+        for (String iri : iris) {
+            try {
+                if (!tsClient.checkDataHasTimeSeries(iri, rdbClient.getConnection())) {
+                    return true;
+                }
+                // If central RDB lookup table ("dbTable") has not been initialised, the time series does not exist
+            } catch (DataAccessException e) {
+                if (e.getMessage().contains("ERROR: relation \"dbTable\" does not exist")) {
+                    return true;
+                } else {
+                    throw e;
+                }
+            } finally {
+                // Always close connection
+                rdbClient.getConnection().close();
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Updates the database with new readings.
+     *
+     * @param excelReadings Excel readings parsed from the Excel Workbook
+     * @param iriMappings   Mappings between measures' names and their corresponding data IRI.
+     */
+    protected void updateData(Map<String, List<?>> excelReadings, Map<String, String> iriMappings) {
+        TimeSeries<Instant> timeSeries = convertReadingsToTimeSeries(excelReadings, iriMappings);
+        // Update each time series
+        try {
+            tsClient.addTimeSeriesData(timeSeries, rdbClient.getConnection());
```

Comment:
This line is most likely to leave an open connection dangling. There also seem to be similar ones elsewhere.